### ISSUE TITLE
Speed up groupKFold 

### ISSUE THREAD

**Issue by mayer79** [topepo/caret#1102]
Hello. The function `groupKFold` is very important in practical applications. However, it is extremely slow. 

Demo code:

```{R}
X <- data.frame(group = sample(1:1e6, size = 3e6, replace = TRUE))
system.time(out <- caret::groupKFold(X$group, k = 5)) # 53 Seconds
```

Here is a fast version (with slightly different output structure):
```{R}
fastGroupKFold <- function(group, k) {
  g_unique <- unique(group)
  stopifnot(k <= length(g_unique))
  g_folds <- sample(k, size = length(g_unique), replace = TRUE)
  split(seq_along(group), g_folds[match(group, g_unique)])
}
system.time(out <- fastGroupKFold(X$group, k = 5)) # 0.6 Seconds
```

It would be fantastic if you could speed up the current implementation of `groupKFold`.

**Comment by topepo** [comment]
If you would make the output format consistent with what we would need, please feel free to make a PR.

**Comment by topepo** [comment]
closed via #1108
